proposals/nnnn-bitwise-copyable.md

```diff
@@ -0,0 +1,270 @@
+
+[^1]: The term "trivial" is used in [SE-138](0138-unsaferawbufferpointer.md) and [SE-0370](0370-pointer-family-initialization-improvements.md) to refer to types with the property above. The discussion below will explain why certain generic or resilient types that are trivial will not in fact be `BitwiseCopyable`.
+
+## Motivation
+
+Swift can compile generic code into an unspecialized form in which the compiled function receives a value and type information about that value.
+Basic operations are implemented by the compiler as calls to a table of "value witness functions."
+
+This approach is flexible, but can represent significant overhead.
+For example, using this approach to copy a buffer with a large number of `Int` values requires a function call for each value.
+
+Constraining the types in generic functions to `BitwiseCopyable` allows the compiler (and in some cases, the developer) to instead use highly efficient direct memory operations in such cases.
+
+The standard library already contains many examples of functions that could benefit from such a concept, and more are being proposed:
+
+The `UnsafeMutablePointer.initialize(to:count:)` function introduced in [SE-0370](0370-pointer-family-initialization-improvements.md) could use a bulk memory copy whenever it statically knew that its argument was `BitwiseCopyable`.
+
+The proposal for [`StorageView`](nnnn-safe-shared-contiguous-storage.md) includes the ability to copy items to or from potentially-unaligned storage, which requires that it be safe to use bulk memory operations:
+```swift
+public func loadUnaligned<T: BitwiseCopyable>(
+  fromByteOffset: Int = 0, as: T.Type
+) -> T
+
+public func loadUnaligned<T: BitwiseCopyable>(
+  from index: Index, as: T.Type
+) -> T
+```
+
+## Proposed solution
+
+We add a new protocol `BitwiseCopyable` to the standard library:
+```swift
+@_marker public protocol BitwiseCopyable {}
+```
+
+Many basic types in the standard library will conformed to this protocol.
+
+Developer's own types may be conformed to the protocol, as well.
+The compiler will check any such conformance and emit a diagnostic if the type contains elements that are not `BitwiseCopyable`.
+
+Furthermore, when building a module, the compiler will infer conformance to `BitwiseCopyable` for any non-resilient struct or enum defined within the module whose stored members are all `BitwiseCopyable`.
+
+Developers cannot conform types defined in other modules to the protocol.
+
+## Detailed design
+
+Our design first conforms a number of core types to `BitwiseCopyable`, and then extends that to aggregate types.
+
+### Standard library changes
+
+Many types and a few key protocols are constrained to `BitwiseCopyable`.
+A few highlights:
+
+* Integer types
+* Floating point types
+* SIMD types
+* Pointer types
+* `Unmanaged`
+* `Optional`
+
+For an exhaustive list, see the [appendix](#all-stdlib-conformers).
+
+### Additional BitwiseCopyable types
```

Comment:
During discussion, a couple other cases came up that we think would be worth calling out explicitly:

* Do raw-value `enum`s get `BitwiseCopyable` inferred if its raw value type is `BitwiseCopyable`? I think there are a couple possible answers here: either "yes", if it would be inferred for the same enum without raw values or associated values since it's a simple ordinal, or "no" if it wouldn't also be inferred for that case. (That is, having a raw value doesn't affect the Swift layout of the enum, so there should be no difference in `BitwiseCopyable` inference from the non-raw-value case.)

* Do imported C structs get `BitwiseCopyable`? If not, is there a way to add it (either retroactively, or via API Notes or `__swift_attr__` in the header)?

Reply:
Added a footnote for raw-value enums.  Inference is based on the enum itself, not the `RawValue` type witness of its synthesized `RawRepresentable` conformance.

Added a section on imported C and C++ types.  Conformance is inferred unless the fields are not representable in Swift (or in the C++ case if the type is non-trivial).  Conformance can be forced even when not inferred via `__attribute__((__swift_attr__("_BitwiseCopyable")))`.